src/sn_utils.c

```diff
@@ -364,13 +366,87 @@ static int assign_one_ip_addr(n2n_sn_t *sss,
     }
   }
   ipaddr->net_addr = net_id | host_id;
-  ipaddr->net_bitlen = sss->dhcp_addr.net_bitlen;
+  ipaddr->net_bitlen = comm->auto_ip_net.net_bitlen;
 
   traceEvent(TRACE_INFO, "Assign IP %s to tap adapter of edge.", ip_subnet_to_str(ip_bit_str, ipaddr));
   return 0;
 }
 
 
+/** checks if a certain sub-network is still available, i.e. does not cut any other community's sub-network */
+int subnet_available(n2n_sn_t *sss,
+                     struct sn_community *comm,
+                     uint32_t net_id,
+                     uint32_t mask) {
+
+  struct sn_community *cmn, *tmpCmn;
+  int success = 1;
+
+  HASH_ITER(hh, sss->communities, cmn, tmpCmn) {
+    if (cmn == comm) continue;
+    if( (net_id <= (cmn->auto_ip_net.net_addr + ~bitlen2mask(cmn->auto_ip_net.net_bitlen)))
+      &&(net_id + ~mask >= cmn->auto_ip_net.net_addr) ) {
+        success = 0;
+        break;
+    }
+  }
+
+  return success;
+}
+
+
+/** The IP address range (subnet) assigned to the community by the auto ip address function of sn. */
+int assign_one_ip_subnet(n2n_sn_t *sss,
+                         struct sn_community *comm) {
+
+  uint32_t net_id, net_id_i, mask, net_increment;
+  uint32_t no_subnets;
+  uint8_t success;
+  in_addr_t net;
+
+  mask = bitlen2mask(sss->min_auto_ip_net.net_bitlen);
+  // number of possible sub-networks
+  no_subnets   = (sss->max_auto_ip_net.net_addr - sss->min_auto_ip_net.net_addr);
+  no_subnets >>= (32 - sss->min_auto_ip_net.net_bitlen);
+  no_subnets  += 1;
+
+  // proposal for sub-network to choose
+  net_id  = pearson_hash_32(comm->community, N2N_COMMUNITY_SIZE) % no_subnets;
+  net_id  = sss->min_auto_ip_net.net_addr + (net_id << (32 - sss->min_auto_ip_net.net_bitlen));
```

Comment:
The community name gets hashed. We need to map the hash value to a sub-network.

To do so, we map the hash to a number of the generally available sub-networks (division modulo) which goes into `net_id` for now.

This number of a sub-network now needs to get mapped to a sub-network itself. It happens in the `(net_id << (32 - sss->min_auto_ip_net.net_bitlen)` part: `32 - bitlen` gives the number of bits used at the end of an IP address for the nodes. Shifting the (still: **number** of sub-network) `net_id` left by this amount of bits, gives a sub-network.

To make this subnet start at the minimum sub-network base address, we have to add just this minimum sub-network base address.

Afterwards, that is later in the code, it checks if this range is still available and if not, it searches downwards and upwards starting from there.

Why do you think it is problematic? It should very well stay in the allowed / defined ranges and give a higher probability that a community finds back to the same IP sub-network.

Reply:
To map the subnets according to the hash value of the community name, I think this method is not very meaningful, because the repetition rate is too high, and ultimately rely on search, which reduces the efficiency. It is better to increment the subnets in the subnet pool in turn .

Reply:
The hash makes kind of sure that it is `not` repeating so easily, so it keeps repetition rate down. Mapping all the possible community names to the 32,768 sub-networks we offer by default, I think there is a low chance for collision – that is a exactly hashes purpose.

The search is required only in case a collision happens. Starting from the hash-based proposal, it should not take longer than the number of currently used sub-networks (worst-case) to find an available network. It probably would be the same if you started from 0 using an incremental assignment. You also would need to search the pool from 0 to find the next available network (you cannot just increase leaving holes for communities that were abandoned).

Big advantage of hashing the community name: You probably get back your old sub-network even when all nodes from a community are gone for a while or longer. That is something that will not happen with incremental pool or pseudo-random assignment.

Reply:
>Big advantage of hashing the community name: You probably get back your old sub-network even when all nodes from a community are gone for a while or longer. That is something that will not happen with incremental pool or pseudo-random assignment.

In this respect, it may indeed be very useful.